src/shims/unix/wasi/foreign_items.rs

```diff
@@ -0,0 +1,187 @@
+
+                    let res = this.write(fd, buf, count.try_into().unwrap()).unwrap();
+                    if res == -1 {
+                        this.write_scalar(Scalar::from_i32(-1), dest)?;
+                        return Ok(EmulateByNameResult::NeedsJumping);
+                    }
+                    total = total.checked_add(res).unwrap();
+                }
+
+                let rp0 = this.addr_to_mplace_ty(rp0, this.tcx.mk_mut_ptr(this.tcx.types.usize))?;
+                this.write_scalar(Scalar::from_i32(total.try_into().unwrap()), &rp0.into())?;
+                this.write_scalar(Scalar::from_u32(0), dest)?;
+            }
+
+            "random_get" => {
+                let [ptr, len] = this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
+
+                let buf = this.addr_to_mplace_ty(ptr, this.tcx.types.u32)?;
+                let len = this.read_machine_usize(len)?;
+
+                this.gen_random(buf.ptr, len)?;
+
+                // Newer versions of the `wasi` crate return a u32, while
+                // older versions return a i16. We support both.
+                let res = match dest.layout.ty {
+                    ty if ty == this.tcx.types.i32 => Scalar::from_i32(0),
+                    ty if ty == this.tcx.types.u16 => Scalar::from_u16(0),
+                    ty => panic!("Unsupported random_get return type: {ty:?}"),
+                };
+                this.write_scalar(res, dest)?;
+            }
+
+            _ => return Ok(EmulateByNameResult::NotSupported),
+        }
+
+        Ok(EmulateByNameResult::NeedsJumping)
+    }
+
+    fn args_get(
+        &mut self,
+        element_heads: &OpTy<'tcx, Provenance>,
+        buffer: &OpTy<'tcx, Provenance>,
+    ) -> InterpResult<'tcx, ()> {
+        let this = self.eval_context_mut();
+
+        let ptr_u8 = this.tcx.mk_mut_ptr(this.tcx.types.u8);
+        let ptr_ptr_u8 = this.tcx.mk_mut_ptr(ptr_u8);
+
+        let ptr_u8_layout = this.layout_of(ptr_u8).unwrap();
+
+        let element_heads = this.addr_to_mplace_ty(element_heads, ptr_ptr_u8)?;
+        let buffer = this.addr_to_mplace_ty(buffer, ptr_u8)?;
+
+        // Based on
+        // https://github.com/bytecodealliance/wasmtime/blob/f85e3f85170bd2c8378977879f4fdf2c6b64de55/crates/wasi-common/src/string_array.rs#L48
+        let mut cursor: u32 = 0;
+        // Can't iterative directly over it to due borrowcheck conflict with 'this'
+        for i in 0..this.machine.args.len() {
+            let arg = &this.machine.args[i];
+            let bytes = arg.as_bytes();
+            let len: u32 = bytes.len().try_into().unwrap();
+            // Make space for `0` terminator.
+            let size = u64::try_from(arg.len()).unwrap().checked_add(1).unwrap();
+
+            let buffer_loc = buffer.ptr.offset(Size::from_bytes(cursor), this)?;
+
+            let arg = OsString::from(arg.to_owned());
+
+            this.write_os_str_to_c_str(&arg, buffer_loc, size)?;
```

Comment:
`size` here is meant to be the size of the target buffer. It makes very little sense to compute it from `arg`, then the size check will obviously always succeed.

How does this wasi API work, how is the buffer size handled? Give how non-standard of a platform wasi is, it would be good to add links to where these functions are documented, so that people that never saw a line of wasi code in their life have any chance of making sense of this.